### PR TITLE
Make JIT attributes t_ and ts_ store Variable instead of Tensor

### PR DESCRIPTION
**Pull request by yf225**
Discussed with @zdevito and we want to use Variable (with `set_requires_grad(false)`) instead of Tensor in all parts of JIT, to eliminate the distinction and the conceptual overhead when trying to figure out which one to use.

This also helps with the Variable/Tensor merge work tracked at https://github.com/pytorch/pytorch/issues/13638, which will make common functions (such as `numel()` / `sizes()` / `dim()`) on Variable much faster when finished.

### CODE REVIEW

File: torch/csrc/jit/constants.cpp [hunk 1]
```diff
@@ -127,7 +129,7 @@ RegisterOperators reg({
           } else if (type->isSubtypeOf(ListType::ofTensors())) {
             const auto& ts = fmap(
                 node->ts(attr::value), [](const at::Tensor& t) -> at::Tensor {
-                  return autograd::make_variable(t);
+                  return const_cast<at::Tensor&>(t).set_requires_grad(false);
```
Comment (zdevito) [inline torch/csrc/jit/constants.cpp]: There is no need to do anything here anymore, just return the Tensors as-is.
```
const auto& ts = node->ts(attr:value):
return [ts](Stack& stack) {
 ...
```

File: torch/csrc/jit/constants.cpp [hunk 2]
```diff
@@ -47,7 +49,7 @@ Value* insertConstant(
     n->ts_(
         attr::value,
         fmap(val.toTensorList()->elements(), [](const at::Tensor& t) {
-          return autograd::Variable(t).data();
+          return const_cast<at::Tensor&>(t).set_requires_grad(false);
```
Comment (zdevito) [inline torch/csrc/jit/constants.cpp]: This should be an assertion that `t` does not require grad. We should not be passing grad recording tensors here.

File: torch/csrc/jit/constants.cpp [hunk 3]
```diff
@@ -20,8 +20,10 @@ Value* insertConstant(
       n->destroy();
       return g.insertNode(g.createUndefined())->output();
     }
-    if (ref.is_variable()) {
-      ref = autograd::Variable(ref).data();
+    if (!ref.is_variable()) {
```
Comment (zdevito) [inline torch/csrc/jit/constants.cpp]: I'd prefer we just `AT_CHECK(ref.is_variable())` and fix the places that pass something else, but this is ok for now if you put the rest in a follow up patch.

File: torch/csrc/jit/constants.cpp [hunk 4]
```diff
@@ -20,8 +20,10 @@ Value* insertConstant(
...
+      ref = autograd::make_variable(ref, /*requires_grad=*/false);
+    } else {
+      ref.set_requires_grad(false);
```
Comment (zdevito) [inline torch/csrc/jit/constants.cpp]: no one should be passing a requires_grad tensor here, so this should be an assert as well. I don't want to silently fix places like this, because it covers up larger logic errors in how constants are created. If it requires_grad it is also likely someone else might mutate it, which would be bad as well.

File: torch/csrc/jit/passes/onnx/peephole.cpp [hunk 5]
```diff
@@ -388,20 +388,20 @@ void fixDefaultRNNState(Graph* graph, Node* n, int input_index) {
   hidden_size->insertBefore(n);
   hidden_size->t_(
       attr::value,
-      at::full(
+      autograd::make_variable(at::full(
```
Comment (zdevito) [inline torch/csrc/jit/passes/onnx/peephole.cpp]: why not just torch::full?

File: torch/csrc/jit/passes/onnx/peephole.cpp [hunk 6]
```diff
@@ -372,7 +372,7 @@ void fixDefaultRNNState(Graph* graph, Node* n, int input_index) {
 
   Node* gather_indices = graph->create(onnx::Constant, 1);
   gather_indices->insertBefore(n);
-  gather_indices->t_(attr::value, at::scalar_to_tensor(at::Scalar(1)));
+  gather_indices->t_(attr::value, autograd::make_variable(at::scalar_to_tensor(at::Scalar(1))));
```
Comment (zdevito) [inline torch/csrc/jit/passes/onnx/peephole.cpp]: there must be a `torch::` version that does the same thing right?

**Review by zdevito — APPROVED** [review]
Some notes in line, otherwise this looks good, pending CI.

File: torch/csrc/jit/constants.cpp (hunk 3 above)
Comment (yf225) [reply torch/csrc/jit/constants.cpp]: Added comment in code:
```
// TODO: fix all cases where we are not passing in a variable,
// and then change this to an AT_ASSERT
```

File: torch/csrc/jit/constants.cpp (hunk 4 above)
Comment (zdevito) [reply torch/csrc/jit/constants.cpp]: Actually, I think this test is bogus. We should not be silently turning parameters into constants. I think what should happen here is a nice user-facing error that says that code is attempting to use a gradient recording tensor as a constant in a trace, and that the user likely wanted to trace the module rather than a function calling the module. I think it is arguable that we should not be implicitly turning anything into constants, but that is a larger change to consider.

Comment (yf225) [reply torch/csrc/jit/constants.cpp]: @zdevito Got it and I can remove the `test_call_python_mod_from_tracing_fn` test.
The other two failing tests look like the following:
```python
def test_call_script_mod_from_tracing_fn(self):
    class ScriptMod(torch.jit.ScriptModule):
        def __init__(self):
            super(ScriptMod, self).__init__()
            self.param = torch.nn.Parameter(torch.rand(4, 3))

        @torch.jit.script_method
        def forward(self, x):
            return torch.mm(x, self.param)

    sm = ScriptMod()

    @_trace(torch.rand(3, 4))
    def traced_fn(x):
        return sm(x) + 1.0

    self.assertExpected(canonical(traced_fn.graph))
```
```python
def test_call_traced_mod_from_tracing_fn(self):
    class TracedModule(torch.nn.Module):
        def __init__(self):
            super(TracedModule, self).__init__()
            self.param = torch.nn.Parameter(torch.rand(4, 3))

        def forward(self, x):
            return torch.mm(x, self.param)

    tm = torch.jit.trace(TracedModule(), torch.rand(3, 4))

    @_trace(torch.rand(3, 4))
    def traced_fn(x):
        return tm(x) + 1.0

    # Note: the parameter self.param from the Python module is inlined
    # into the graph
    self.assertExpected(canonical(traced_fn.graph))
```
These two tests have the same issue in that `self.param = torch.nn.Parameter(torch.rand(4, 3))` has `requires_grad=True` by default (and if I manually set `self.param` to `requires_grad=False`, the test will pass). Are the use cases in these two tests valid / should we also remove them?

Comment (zdevito) [reply torch/csrc/jit/constants.cpp]: We should keep these tests because they are checking that some interactions between tracing and script work. However, it is fine to explicitly set requires_grad=False to make them pass. That is not what they were testing before.

Comment (zdevito) [reply torch/csrc/jit/constants.cpp]: That might be true of the first test as well. We can just turn of the requires_grad to make them all pass.
